### PR TITLE
Fix erroneous react-dom usage in native tests

### PR DESCRIPTION
**Pull request by dcalhoun**
## Description
Reinstate `react-platform.native.js` file, which was removed in https://github.com/WordPress/gutenberg/commit/b7b62d2d3db4395ce5d2ad42bd1fc0f1c3c2f12d, as it was no longer strictly necessary. However, the removal led to Jest loading the DOM-specific version of the file when mocking Gutenberg modules that depended upon `react-platform.js`. Jest would seemingly load `react-dom` when attempting to auto-mock the module, e.g. [`reusable-block-types-tab.native.js`](https://github.com/WordPress/gutenberg/blob/bea24fbd7ec829ed84c5dd30e5416052440f5171/packages/block-editor/src/components/inserter/test/reusable-blocks-tab.native.js#L19) > [`use-select/index.js`](https://github.com/WordPress/gutenberg/blob/bea24fbd7ec829ed84c5dd30e5416052440f5171/packages/data/src/components/use-select/index.js#L10) > [`@wordpress/element`](https://github.com/WordPress/gutenberg/blob/bea24fbd7ec829ed84c5dd30e5416052440f5171/packages/element/src/index.js#L3) > [`react-platform.js`](https://github.com/WordPress/gutenberg/blob/bea24fbd7ec829ed84c5dd30e5416052440f5171/packages/element/src/react-platform.js#L4-L9). Reinstating `react-platform.native.js` addresses this issue by ensuring that Jest does not encounter an import of `react-dom`.

<details><summary>react-dom Error Details</summary>

```
TypeError: Cannot use 'in' operator to search for 'WebkitAnimation' in undefined

  at getVendorPrefixedEventName (node_modules/react-dom/cjs/react-dom.development.js:5011:58)
  at node_modules/react-dom/cjs/react-dom.development.js:5019:21
  at Object.<anonymous> (node_modules/react-dom/cjs/react-dom.development.js:26261:5)
  at Object.<anonymous> (node_modules/react-dom/index.js:37:20)
```

</details>

Additionally, these changes fix a false-positive `ReactNativeEditor` test. This test failed when run in isolation. It would appear it was dependent upon the `jest.mock('../setup')` found in sibling tests.

After defining a mock for this specific test, it was discovered that the assertion did not await the asynchronous query found within the test. The assertion resulted in a false-positive as the returned query `Promise` technically matches `toBeDefined`. `async`/`await` was added to ensure the test assertion awaits the query result, as well as fixes the following related log warning.

<details><summary>Async Test Warning</summary>

```
A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks. Active timers can also cause this, ensure that .unref() was called on them.
```

</details>

## How has this been tested?
- Verified native tests pass locally.
- Verified native Demo editor launched on device without issue.

## Screenshots 
n/a

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [x] I've updated related schemas if appropriate.

### CODE REVIEW

File: packages/react-native-editor/src/test/index.test.js
Comment (dcalhoun) [inline packages/react-native-editor/src/test/index.test.js:186]: This test failed when run in isolation, as it appeared dependent upon the mock of `../setup` found in sibling tests.

File: packages/react-native-editor/src/test/index.test.js
Comment (dcalhoun) [inline packages/react-native-editor/src/test/index.test.js:189]: The missing `await` for the asynchronous query meant the test assertion succeeded as a false-positive.

Comment (fluiddot) [reply packages/react-native-editor/src/test/index.test.js:189]: Oh sorry for that, I fell into the same trap when adding other integration tests in the past 🤦 . Thanks for catching it 🙇 !

I did an identical test in GB-mobile ([reference](https://github.com/wordpress-mobile/gutenberg-mobile/blob/c847b87cec6144ab6787ab837302ec87266c7eec/src/test/index.js#L57-L64)), so I'll apply the same fix there 🔧 .

File: packages/react-native-editor/src/test/index.test.js
Comment (fluiddot) [reply packages/react-native-editor/src/test/index.test.js:186]: That's interesting, my idea for this test case was actually to verify that the entire editor is initialized including rendering the [initial HTML](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/src/initial-html.js). In any case, and in order to prevent blocking the PR and Jest upgrade, we could stick with this change and tackle my original idea in a future PR 👍 .

**Comment by gziolo** [comment]
The override for `react-platform.js` file makes a lot of sense because there is no `react-dom` in React Native 👍🏻 

Let's wait for approval from someone that is better suited to confirm it's the right approach. CI is finally green so I like that PR a lot 😄

**Review by fluiddot — APPROVED** [review]
LGTM 🎊 !

I've just left a couple of comments but none of them should block the PR.

> Reinstate react-platform.native.js file, which was removed in b7b62d2, as it was no longer strictly necessary. However, the removal led to Jest loading the DOM-specific version of the file when mocking Gutenberg modules that depended upon react-platform.js. Jest would seemingly load react-dom when attempting to auto-mock the module, e.g. reusable-block-types-tab.native.js > use-select/index.js > @wordpress/element > react-platform.js. Reinstating react-platform.native.js addresses this issue by ensuring that Jest does not encounter an import of react-dom.

Great catch, I didn't realize about this issue when I did the initialization refactor 🤦 , thanks for spotting it 🙇 . I also noticed that due to this, we included `react-dom` in the JS bundle (checked in the [latest source map](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/bundle/ios/App.js.map)), but I guess is innocuous apart from increasing the bundle's size a bit (the size of `react-dom.production.min.js` is 100kb).

File: packages/react-native-editor/src/test/index.test.js
Comment (dcalhoun) [reply packages/react-native-editor/src/test/index.test.js:186]: @fluiddot I leveraged a mock of `../setup` as a quick fix to avoid errors from additional dependencies of the `../setup` component. I think asserting against a deeper render makes a lot of sense. 👍🏻

Comment (fluiddot) [reply packages/react-native-editor/src/test/index.test.js:186]: > In any case, and in order to prevent blocking the PR and Jest upgrade, we could stick with this change and tackle my original idea in a future PR 👍 .

@dcalhoun heads up that I created [a PR](https://github.com/WordPress/gutenberg/pull/37955) related to this comment.

**Pull request merged**
